lib/client.js

```diff
@@ -139,7 +163,23 @@ module.exports = function (dependencies) {
           let error = new VError("stream ended unexpectedly");
           resolve({ device, error });
         }
-      })
+      });
+
+      request.setTimeout(this.config.requestTimeout, () => {
```

Comment:
1. If the remote server stops sending any TCP/UDP packets back (e.g. no resets, no acknowledgements, no close notifications), then I'm concerned that this could resolve promises but never close the http/2 session if requests are sent but no requests are getting responses. (on('error') would be emitted in response to a packet)

   (I assume the failure mode I've been seeing is similar to this, which is what I wrote my PR to address, including tracking the most recent successsful response time across *all* requests, to work even if bandwidth was limited)
2. If this doesn't close the connection, it would leak memory.

Reply:
1 - So what's your suggestion here? Close and destroy the session to create a new one? I'm not sure if it is necessary since the server can timeout for this request but can work for the next one. I did a stress test and saw this happening. One request timed out but the following ones went through. I believe that if something wrong goes with the whole session, one of the other events will be fired.
2 - You're right. I will close the request.

Reply:
2 - done

Reply:
> 1 - So what's your suggestion here? Close and destroy the session to create a new one? I'm not sure if it is necessary since the server can timeout for this request but can work for the next one. I did a stress test and saw this happening. One request timed out but the following ones went through. I believe that if something wrong goes with the whole session, one of the other events will be fired.

The suggestion was to track _lastSuccessfulApnResponseMillis - if there are active requests, only give up on the connection if the most recently received successful response was received earlier than `now - timeout`.

I can't say more without being sure of what the stress test was doing and what it was tested against

Reply:
I'm not sure if we should do that in this PR. It brings an extra complexity to the code. I believe we could address on a separate PR if we continue having issues.